src/components/tl/ucp/allgather/allgather_bruck.c

```diff
@@ -0,0 +1,254 @@
+    void        *tmprecv, *tmpsend;
+
+    if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+        return;
+    }
+
+    /* On each step doubles distance */
+    distance = 1 << task->tagged.send_posted;
+    tmpsend  = rbuf;
+    while (distance < tsize) {
+
+        recvfrom = (trank + distance) % tsize;
+        sendto   = (trank + tsize - distance) % tsize;
+
+        tmprecv = PTR_OFFSET(tmpsend, distance * data_size);
+
+        if (distance <= tsize >> 1) {
+            blockcount = distance;
+        } else {
+            /* send-recv all reminder*/
+            blockcount = tsize - distance;
+        }
+
+        /* Sendreceive */
+        UCPCHECK_GOTO(ucc_tl_ucp_send_nb(tmpsend, blockcount * data_size, rmem,
+                                         sendto, team, task),
+                      task, out);
+        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(tmprecv, blockcount * data_size, rmem,
+                                         recvfrom, team, task),
+                      task, out);
+
+        if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+            return;
+        }
+
+        distance = 1 << task->tagged.send_posted;
+    }
+
+    /* post processing step */
+    if (trank != 0) {
+        if (UCC_MEMORY_TYPE_HOST == rmem) {
+            // copy blocks [0 .. (size - rank - 1)] from rbuf to shift buffer
+            status = ucc_mc_memcpy(scratch_header->addr, rbuf, scratch_size,
+                                   UCC_MEMORY_TYPE_HOST, rmem);
+            if (ucc_unlikely(status != UCC_OK)) {
+                tl_error(UCC_TASK_LIB(task),
+                         "failed to copy data to scratch buffer");
+                ucc_tl_ucp_coll_finalize(&task->super);
+                return;
+            }
+            // move blocks [(size - rank) .. size] from rbuf to beginning of rbuf
+            // TODO: rewrite to cycle to get rid of overlap
+            memmove(rbuf, PTR_OFFSET(rbuf, scratch_size), trank * data_size);
+            // copy blocks from shift buffer starting at block [rank] in rbuf.
+            status = ucc_mc_memcpy(PTR_OFFSET(rbuf, trank * data_size),
+                                   scratch_header->addr, scratch_size, rmem,
+                                   UCC_MEMORY_TYPE_HOST);
+            if (ucc_unlikely(status != UCC_OK)) {
+                tl_error(UCC_TASK_LIB(task),
+                         "failed to copy data from scratch to rbuff buffer");
+                ucc_tl_ucp_coll_finalize(&task->super);
+                return;
+            }
+        } else {
+            /* In case of non host memory we perform two copy to host buffer and then back to device, 3 memcopy in total */
+            /* TODO: replace with generic kernel to do bruck post step in sinle launch on device */
+            status = ucc_mc_memcpy(
+                PTR_OFFSET(scratch_header->addr, trank * data_size), rbuf,
+                (tsize - trank) * data_size, UCC_MEMORY_TYPE_HOST, rmem);
```

Comment:
in case of device side buffer we allocate scratch buffer on host memory, but size of scratch = tsize * data_size, so full buffer, not part as in case of host memory case.

Reply:
oh, I see, nevermind